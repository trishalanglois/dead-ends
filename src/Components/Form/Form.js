import React, { Component } from 'react';
import './Form.scss';


class Form extends Component {
  constructor() {
    super();
    this.state = {
      location: ''
    }
  }


  render() {
    return (
      <p> this is my form component eeeeeee </p>

    )
  }
}

export default Form;
