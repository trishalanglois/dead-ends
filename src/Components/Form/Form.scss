@import '../../scss/variables';

.form-container {
  background-image: url('../../images/background.jpg');
  background-size: contain;
  background-position: center;
  height: 88vh;
  width: 100vw;
}
