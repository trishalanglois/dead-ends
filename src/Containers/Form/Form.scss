@import '../../scss/variables';

.form-container {
  background-image: $background-image;
  background-size: contain;
  background-position: center;
  height: 88vh;
  width: 100vw;
  text-align: center;
}

form {
  display: flex;
  flex-direction: column;
  width: 20em;
  align-items: center;
  background-color: $dark-color;
  border: 2px solid $medium-grey;
  position: absolute;
  z-index: 2;
  top: 18em;
  left: 43%;
  padding: 1em;
}


.input-label {
  color: $light-grey;
  font-size: 1.75em;
  font-family: $accent-font;
  margin: .25em;
}

.form-input {
  font-size: 1.5em;
  font-family: $main-font;
  width: 12em;
  margin-bottom: .75em;
  padding: .2em;
    ::placeholder {
      color: $medium-grey;
      font-style: italic;
      font-weight: 300;
    }
}

.form-button {
  font-size: 1.5em;
  margin: 1em;
  width: 12em;
  font-family: $accent-font;
  background-color: $dark-grey;
  border-style: none;
  padding: .25em;
  &:hover {
    cursor: pointer;
    transform: scale(1.03);
    background-color: $medium-grey;
  }
}

.error {
  color: $favorite-color;
  font-family: $main-font;
}
