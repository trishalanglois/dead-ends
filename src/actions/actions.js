export const currentUser = user => ({
  type: 'SET_USER',
  name: user
})

export const setLocation = location => ({
  type: 'SET_LOCATION',
  location
})

export const setSpookyLocations = locations => ({
  type: 'SET_SPOOKY_LOCATIONS',
  locations
})

export const updateFavorites = location => ({
  type: 'UPDATE_FAVORITES',
  location
})

export const updateClickedCard = location => ({
  type: 'UPDATE_CLICKED_CARD',
  location
})
