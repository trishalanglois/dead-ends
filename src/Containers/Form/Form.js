import React, { Component } from 'react';
import './Form.scss';
import { connect } from 'react-redux';
import { currentUser, setLocation, setSpookyLocations } from '../../actions/actions';
import locationData from '../../data.json';
import { Redirect, Link } from 'react-router-dom';


export class Form extends Component {
  constructor() {
    super();
    this.state = {
      username: '',
      location: '',
      error: false,
      ready: false,
      noLocations: false
    }
  }

  handleChange = (e) => {
    this.setState({ [e.target.name]: e.target.value })
  }

  handleLocationSubmit = (e) => {
    e.preventDefault();
    this.checkLocation(this.state.location);
  }

  checkLocation = (location) => {
    const regex = /([A-Za-z]+(?: [A-Za-z]+)*),? ([A-Za-z]{2})/;
    if (location.match(regex)) {
      this.props.updateUser(this.state.username);
      this.props.setLocation(this.state.location);
      this.findLocations(this.state.location);
    } else {
      this.handleError();
    }
  }

  findLocations = (location) => {
    const splitLocation = location.split(', ');
    const city = splitLocation[0];
    const state = splitLocation[1];
    const filteredLocations = locationData.filter(spookySpot => {
      return spookySpot.city === city && spookySpot.state_abbrev === state
    })
    if (!filteredLocations.length) {
      this.setState({noLocations: true})
    } else {
      this.setState({ ready: true })
      this.props.setSpookyLocations(filteredLocations);
    }
  }

  handleError() {
    this.setState({error: true})
  }

  render() {
    if (this.state.ready) {
      return <Redirect to='/locations' />
    }
    return (
      <main className="form-container">
        <form>
          <label className="input-label">Ghost Hunter Name</label>
          <input
            className="form-input"
            value={this.state.username}
            name="username"
            type="text"
            onChange={(e) => this.handleChange(e)}
          />
          <label className="input-label">Location</label>
          <input
            className="form-input"
            placeholder="e.g. Denver, CO"
            value={this.state.location}
            name="location"
            type="text"
            onChange={(e) => this.handleChange(e)}
          />
          {this.state.error &&
            <p className="error">Please enter a valid city and state.</p>
          }
          {this.state.noLocations && <p className="error">This location is not haunted.  Please enter a spookier location.</p>}
            <button
              className="form-button"
              type="button"
              onClick={(e) => this.handleLocationSubmit(e)}
            >
              Begin the Hunt
            </button>
        </form>
      </main>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  updateUser: user => dispatch(currentUser(user)),
  setLocation: location => dispatch(setLocation(location)),
  setSpookyLocations: locations => dispatch(setSpookyLocations(locations))
})

export default connect(null, mapDispatchToProps)(Form);
