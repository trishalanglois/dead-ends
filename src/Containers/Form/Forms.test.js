import React from 'react';
import { Form, mapDispatchToProps } from './Form';
import { shallow } from 'enzyme';
import { currentUser, setLocation, setSpookyLocations } from '../../actions/actions';

describe('Form', () => {
  let wrapper, mockEvent, mockCurrentUser, mockSetLocation, mockSetSpookyLocations;
  describe('Form Component', () => {
    beforeEach(() => {
      mockEvent = { target: { name: 'username', value: 'KWoo'}}
      mockCurrentUser = jest.fn();
      mockSetLocation = jest.fn();
      mockSetSpookyLocations = jest.fn();
      wrapper = shallow(<Form
        updateUser={mockCurrentUser}
        setLocation={mockSetLocation}
        setSpookyLocations={mockSetSpookyLocations}
      />)
    })

    it('should match the snapshot when there is no error in state', () => {
      expect(wrapper).toMatchSnapshot();
    })

    it('should match the snapshot when there is an error in state', () => {
      wrapper.setState({error: true})
      expect(wrapper).toMatchSnapshot();
    })

    it('should set the state when handleChange is called', () => {
      wrapper.setState({ username: '', location: '', error: false });
      wrapper.instance().handleChange(mockEvent);

      expect(wrapper.state()).toEqual({ username: 'KWoo', location: '', error: false, ready: false, noLocations: false })
    })

    it('should call handleChange when the name input is changed', () => {
      wrapper.instance().handleChange = jest.fn();
      const mockNameEvent = { target: { username: 'username', value: 'KWoo'}}

      wrapper.find('.name-input').simulate('change', mockNameEvent);

      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockNameEvent);
    });

    it('should call handleChange when the location input is changed', () => {
      wrapper.instance().handleChange = jest.fn();
      const mockLocationEvent = { target: { location: 'location', value: 'Denver, CO'}}

      wrapper.find('.location-input').simulate('change', mockLocationEvent);

      expect(wrapper.instance().handleChange).toHaveBeenCalledWith(mockLocationEvent);
    });

    it('should call checkLocation with a location when handleLocationSubmit is called', () => {
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.instance().checkLocation = jest.fn();
      wrapper.setState({ location: 'Denver, CO'});
      wrapper.find('.form-button').simulate('click', mockEvent);

      expect(wrapper.instance().checkLocation).toHaveBeenCalledWith('Denver, CO')
    })

    describe('checkLocation', () => {
      it('should call mapDispatchToProps with a user when a user enters a location', () => {
        const mockMapDispatchToProps = jest.fn();
        const actionToDispatch = currentUser('KWoo');
        const mockProps = mapDispatchToProps(mockMapDispatchToProps);
        mockProps.updateUser('KWoo');

        expect(mockMapDispatchToProps).toHaveBeenCalledWith(actionToDispatch);
      })
    })

    it('should call mapDispatchToProps with a location when the user enters a location', () => {
      const mockMapDispatchToProps = jest.fn();
      const actionToDispatch = setLocation('Denver, CO');
      const mockProps = mapDispatchToProps(mockMapDispatchToProps);
      mockProps.setLocation('Denver, CO');

      expect(mockMapDispatchToProps).toHaveBeenCalledWith(actionToDispatch);
    })

    it('should call mapDispatchToProps with an array of locations after a user has entered a location', () => {
      const mockMapDispatchToProps = jest.fn();
      const actionToDispatch = setSpookyLocations([
          {
            "city": "Denver",
            "country": "United States",
            "description": "BOO!",
            "location": "Ada Cemetery",
            "state": "Colorado",
            "state_abbrev": "CO",
            "longitude": -85.50489309999999,
            "latitude": 42.9621061,
            "city_longitude": -85.4954803,
            "city_latitude": 42.960727
          },
          {
            "city": "Denver",
            "country": "United States",
            "description": "Spooky stuff.",
            "location": "North Adams Rd.",
            "state": "Colorado",
            "state_abbrev": "CO",
            "longitude": -84.3818434,
            "latitude": 41.9714248,
            "city_longitude": -84.3471684,
            "city_latitude": 41.986434
        }]);
      const mockProps = mapDispatchToProps(mockMapDispatchToProps);
      mockProps.setSpookyLocations([
          {
            "city": "Denver",
            "country": "United States",
            "description": "BOO!",
            "location": "Ada Cemetery",
            "state": "Colorado",
            "state_abbrev": "CO",
            "longitude": -85.50489309999999,
            "latitude": 42.9621061,
            "city_longitude": -85.4954803,
            "city_latitude": 42.960727
          },
          {
            "city": "Denver",
            "country": "United States",
            "description": "Spooky stuff.",
            "location": "North Adams Rd.",
            "state": "Colorado",
            "state_abbrev": "CO",
            "longitude": -84.3818434,
            "latitude": 41.9714248,
            "city_longitude": -84.3471684,
            "city_latitude": 41.986434
        }])
      expect(mockMapDispatchToProps).toHaveBeenCalledWith(actionToDispatch)
    });

    it('should call handleError if the user enters an incorrect location', () => {
      wrapper.instance().handleError = jest.fn();
      wrapper.instance().checkLocation('123');

      expect(wrapper.instance().handleError).toHaveBeenCalled();
    })

    describe('findLocations', () => {
      it('should set noLocations state to true if there are no matching locations from the user', () => {
        const filteredLocations = [];
        const mockEvent = wrapper.instance().findLocations('');

        expect(wrapper.state('noLocations')).toEqual(true);
      })

      // it('should state the ready state to true and call setSpookyLocations if there are matching locations from the user', () => {
      //   const filteredLocations = [{
      //     "city": "Denver",
      //     "country": "United States",
      //     "description": "BOO!",
      //     "location": "Ada Cemetery",
      //     "state": "Colorado",
      //     "state_abbrev": "CO",
      //     "longitude": -85.50489309999999,
      //     "latitude": 42.9621061,
      //     "city_longitude": -85.4954803,
      //     "city_latitude": 42.960727
      //   }];
      //   const mockEvent = wrapper.instance().findLocations('Denver, CO');
      //   wrapper.instance().setSpookyLocations = jest.fn();
      //
      //   expect(wrapper.state('ready')).toEqual(true);
      //   expect(wrapper.instance().setSpookyLocations).toHaveBeenCalledWith(filteredLocations)
      // })
    })

    it('should set the error state to true if handleError has been called', () => {
      wrapper.instance().handleError();
      
      expect(wrapper.state('error')).toEqual(true)
    })
  })
})
