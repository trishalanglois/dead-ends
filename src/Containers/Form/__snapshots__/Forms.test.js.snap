// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Form Component should match the snapshot when there is an error in state 1`] = `
<main
  className="form-container"
>
  <form>
    <label
      className="input-label"
    >
      Ghost Hunter Name
    </label>
    <input
      className="name-input form-input"
      name="username"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      className="input-label"
    >
      Location
    </label>
    <input
      className="form-input location-input"
      name="location"
      onChange={[Function]}
      placeholder="e.g. Denver, CO"
      type="text"
      value=""
    />
    <p
      className="error"
    >
      Please enter a valid city and state.
    </p>
    <button
      className="form-button"
      onClick={[Function]}
      type="button"
    >
      Begin the Hunt
    </button>
  </form>
</main>
`;

exports[`Form Form Component should match the snapshot when there is no error in state 1`] = `
<main
  className="form-container"
>
  <form>
    <label
      className="input-label"
    >
      Ghost Hunter Name
    </label>
    <input
      className="name-input form-input"
      name="username"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      className="input-label"
    >
      Location
    </label>
    <input
      className="form-input location-input"
      name="location"
      onChange={[Function]}
      placeholder="e.g. Denver, CO"
      type="text"
      value=""
    />
    <button
      className="form-button"
      onClick={[Function]}
      type="button"
    >
      Begin the Hunt
    </button>
  </form>
</main>
`;
